pyccel/ast/core.py

```diff
@@ -2231,31 +2231,93 @@ def __new__(
         precision=0,
         is_argument=False
         ):
+        return Basic.__new__(cls)
+
+    def __init__(
+        self,
+        dtype,
+        name,
+        rank=0,
+        allocatable=False,
+        is_stack_array = False,
+        is_pointer=False,
+        is_target=False,
+        is_polymorphic=None,
+        is_optional=None,
+        shape=None,
+        cls_base=None,
+        cls_parameters=None,
+        order='C',
+        precision=0,
+        is_argument=False
+        ):
+
+        # ------------ PyccelAstNode Properties ---------------
         if isinstance(dtype, str) or str(dtype) == '*':
 
             dtype = datatype(str(dtype))
         elif not isinstance(dtype, DataType):
             raise TypeError('datatype must be an instance of DataType.')
 
+        if not isinstance(rank, int):
+            raise TypeError('rank must be an instance of int.')
+
+        if rank == 0:
+            shape = ()
+
+        if not precision:
+            if isinstance(dtype, NativeInteger):
+                precision = default_precision['int']
+            elif isinstance(dtype, NativeReal):
+                precision = default_precision['real']
+            elif isinstance(dtype, NativeComplex):
+                precision = default_precision['complex']
+            elif isinstance(dtype, NativeBool):
+                precision = default_precision['bool']
+        if not isinstance(precision,int):
+            raise TypeError('precision must be an integer.')
+
+        self._dtype = dtype
+        self._shape = process_shape(shape)
+        self._rank  = rank
+        self._precision = precision
+
+        # ------------ Variable Properties ---------------
+        # if class attribute
+        if isinstance(name, str):
+            name = name.split(""".""")
+            if len(name) == 1:
+                name = name[0]
+            else:
+                name = DottedName(*name)
+
+        if not isinstance(name, (str, DottedName)):
+            raise TypeError('Expecting a string or DottedName, given {0}'.format(type(name)))
+        self._name = name
+
         if allocatable is None:
             allocatable = False
```

Comment:
I don't know. I also found that a bit odd but it was in the code previously. I think you are right and we should remove it

Reply:
It seems that `_infere_type` passes several values as None:
```
        d_var['allocatable'   ] = None
        d_var['is_stack_array'] = None
        d_var['is_pointer'    ] = None
        d_var['is_target'     ] = None
        d_var['is_polymorphic'] = None
        d_var['is_optional'   ] = None
        d_var['cls_base'      ] = None
        d_var['cls_parameters'] = None
```
This allows the default values to be set in Variable and be easily adjusted

Reply:
We could set the default value to None in the arguments though to avoid specifying it 3 times for each argument (once in the args for `__new__`, once in the args for `__init__`, and once in the body of `__init__` )

Reply:
To avoid some of the repetition we could use `**kwargs` in the signature of `__new__`, and leave the named arguments (with their default values) in the `__init__`.

As for using `None` as a default value, that seems totally useless to me. Why do we even need to add those entries to the `d_var` dictionary?

Reply:
Ah! I thought the signatures had to match. I will do that.

I agree, it seems that the dictionary values are referenced directly. But this also seems strange as it means that the default values haven't been taken into account. @saidctb  do you have any insight here?

Reply:
I would very much like to clarify this point. I am accepting the pull request as it is, but wait for further input from @saidctb before merging.

Reply:
yes maybe we can delete the None value from the dictionary when we create the instance,
or have the default values, but I'm not sure for example in the `allocatable` keyword case, it only makes sense to have it set to `True` or `False` if the `rank>0` otherwise`None` seems more suitable

Reply:
I think `False` makes sense for `rank=0`, it just means it's not a pointer. For example a class instance may or may not be allocatable.
On a similar note, can strings be `allocatable`?

Reply:
I don't think they can , or we never used it before but I might be wrong

Reply:
Any scalar polymorphic quantity can be `allocatable` in Fortran, as well as strings and arrays. So setting `allocatable=False` makes sense for pretty much any variable that is not allocated on the heap.

Reply:
@EmilyBourne Could you remove the unused keys from the dictionary?